### PR TITLE
[WIP] Fixed entry point omitted from the PDB when it's the same as input filename

### PR DESCRIPTION
**Pull request by adam-yang**

### CODE REVIEW

File: tools/clang/tools/dxcompiler/dxcshadersourceinfo.h [hunk 1]
```diff
@@ -56,7 +56,7 @@ struct SourceInfoWriter {
   Buffer m_Buffer;
 
   const hlsl::DxilSourceInfo *GetPart() const;
-  void Write(llvm::StringRef targetProfile, llvm::StringRef entryPoint, clang::CodeGenOptions &cgOpts, clang::SourceManager &srcMgr);
+  void Write(clang::CodeGenOptions &cgOpts, llvm::ArrayRef<hlsl::options::ArgPair> ArgPairs, clang::SourceManager &srcMgr);
```
Comment (damyanp) [inline tools/clang/tools/dxcompiler/dxcshadersourceinfo.h:59]: Any reason this uses the full `hlsl::optons::ArgPair` while other instances in this file use just `options::ArgPair`?

File: lib/DxcSupport/HLSLOptions.cpp [hunk 2]
```diff
@@ -69,6 +69,116 @@ const OptTable * hlsl::options::getHlslOptTable() {
...
+    //   { "",    "Value0" }
+    //   { "",    "Value1" }
+    //   { "",    "Value2" }
+    // 
+    // ..Which is not ideal, since the whole point of arg pair is to easily break down what arg is part of what.
+    // 
+    // Alternative is to try this:
+    //   { "arg", "Value0 Value1 Value2" }
+    // 
+    // But this causes problem when recompiling
```
Comment (damyanp) [inline lib/DxcSupport/HLSLOptions.cpp:140]: > "But this causes problem when recompiling"

What does this mean?  Is the problem that it doesn't round-trip properly - ie you can't trivially reconstruct a command-line when it's stored in this alternate form?

File: tools/clang/tools/dxcompiler/dxcshadersourceinfo.h (hunk 1 above)
Comment (damyanp) [inline tools/clang/tools/dxcompiler/dxcshadersourceinfo.h:59]: Params to this function look like they have inconsistent naming conventions to me:
* `cgOpts`
* `ArgPairs`
* `srcMgr`

File: lib/DxcSupport/HLSLOptions.cpp [hunk 3]
```diff
@@ -69,6 +69,116 @@ const OptTable * hlsl::options::getHlslOptTable() {
   return g_HlslOptTable;
 }
 
+static std::vector<ArgPair> ComputeArgPairsFromArgList(const llvm::opt::InputArgList &argList) {
```
Comment (damyanp) [inline lib/DxcSupport/HLSLOptions.cpp:72]: FWIW, it's a shame there's not a single set of unit tests I can look at that directly exercise all the various cases and edge cases that `ComputeArgPairsFromArgList` handles.  Presumably there's good coverage for all 5 or so types of arguments this supports through the rest of the test suite?

**Review by damyanp — APPROVED** [review]

File: lib/DxcSupport/HLSLOptions.cpp (hunk 3 above)
Comment (adam-yang) [reply lib/DxcSupport/HLSLOptions.cpp:72]: Actually we do in the new test case. -Zi and -Od are ::FlagClass, -T and -E are `arg->getNumValues() == 1`. The input file is `InputClass`.

Comment (adam-yang) [reply lib/DxcSupport/HLSLOptions.cpp:72]: This made me revise this area. Before the latest revision, the OP_INPUT check would skip the iteration and we would never hit the ::InputClass case. Now the skip is done by the InputClass.

File: lib/DxcSupport/HLSLOptions.cpp (hunk 2 above)
Comment (adam-yang) [reply lib/DxcSupport/HLSLOptions.cpp]: Added an explanation for how it actually failed.

File: tools/clang/tools/dxcompiler/dxcshadersourceinfo.h (hunk 1 above)
Comment (adam-yang) [reply tools/clang/tools/dxcompiler/dxcshadersourceinfo.h]: Removed it. That might have been from copypasting from other places.

Comment (adam-yang) [reply tools/clang/tools/dxcompiler/dxcshadersourceinfo.h]: Fixed it.

**Review by damyanp — APPROVED** [review]

File: tools/clang/tools/dxcompiler/dxcompilerobj.cpp [hunk 4]
```diff
@@ -1571,15 +1571,14 @@ class DxcCompiler : public IDxcCompiler3,
     else
       compiler.getCodeGenOpts().HLSLSignaturePackingStrategy = (unsigned)DXIL::PackingStrategy::Default;
 
-    // Constructing vector of wide strings to pass in to codegen. Just passing
-    // in pArguments will expose ownership of memory to both CodeGenOptions and
-    // this caller, which can lead to unexpected behavior.
-    for (UINT32 i = 0; i != Opts.Args.getNumInputArgStrings(); ++i) {
```
[the remark this answers is not in the document]
Comment (adam-yang) [reply tools/clang/tools/dxcompiler/dxcompilerobj.cpp:1577]: I think that's the right short term fix to minimize risk for release, so I'll make a separate PR with just that fix, but I still think this is the right change to do long term. I'll rebase later.

File: include/dxc/Support/HLSLOptions.td [hunk 5]
```diff
@@ -170,6 +170,11 @@ def opt_enable : Separate<["-", "/"], "opt-enable">, Group<hlsloptz_Group>, Flag
 def opt_select : MultiArg<["-", "/"], "opt-select", 2>, MetaVarName<"<opt> <variant>">, Group<hlsloptz_Group>, Flags<[CoreOption, DriverOption, HelpHidden]>,
   HelpText<"Select this optimization variant.">;
 
+def CommaJoinTest : CommaJoined<["-"], "CommaJoinTest=">, Group<spirv_Group>, Flags<[CoreOption, HelpHidden]>,
```
Comment (llvm-beanz) [inline include/dxc/Support/HLSLOptions.td:173]: I assume these options aren’t actually SPIR-V specific right? They should probably be in a different Group,

File: lib/DxcSupport/HLSLOptions.cpp [hunk 6]
```diff
@@ -69,6 +69,119 @@ const OptTable * hlsl::options::getHlslOptTable() {
...
+  std::vector<ArgPair> ret;
```
Comment (llvm-beanz) [inline lib/DxcSupport/HLSLOptions.cpp:73]: nit: `SmallVector<>` is probably better here to reduce resize occurrences.

File: lib/DxcSupport/HLSLOptions.cpp [hunk 7]
```diff
@@ -69,6 +69,119 @@ const OptTable * hlsl::options::getHlslOptTable() {
...
+
+    llvm::SmallString<64> argumentStorage; // Buffer for when we need to concatenate things.
+
+    // The input argument should be excluded from this list. All the uses for
+    // the ArgPair list used to manually filter out the input file. Do it here
+    // in the source instead.
+    if (arg->getOption().getKind() == llvm::opt::Option::InputClass) {
+      continue;
+    }
+    // This is a flag without value, like -Zi, or -Od
```
Comment (llvm-beanz) [inline lib/DxcSupport/HLSLOptions.cpp:87]: nit: can you clang-format the new code?

File: lib/DxcSupport/HLSLOptions.cpp [hunk 8]
```diff
@@ -69,6 +69,119 @@ const OptTable * hlsl::options::getHlslOptTable() {
   return g_HlslOptTable;
 }
 
+static std::vector<ArgPair> ComputeArgPairsFromArgList(const llvm::opt::InputArgList &argList) {
```
Comment (llvm-beanz) [inline lib/DxcSupport/HLSLOptions.cpp:72]: Rather than returning a std::vector which guarantees a copy, can we pass a `SmallVectorImpl<>&` to avoid copying the array on return? That also allows us to avoid resizes.

File: lib/DxcSupport/HLSLOptions.cpp (hunk 3 above)
Comment (llvm-beanz) [reply lib/DxcSupport/HLSLOptions.cpp:72]: Since this function is pretty isolated and static, can we write unit tests for this function?

File: lib/DxcSupport/HLSLOptions.cpp [hunk 9]
```diff
@@ -69,6 +69,119 @@ const OptTable * hlsl::options::getHlslOptTable() {
...
+
+    // The input argument should be excluded from this list. All the uses for
+    // the ArgPair list used to manually filter out the input file. Do it here
+    // in the source instead.
+    if (arg->getOption().getKind() == llvm::opt::Option::InputClass) {
+      continue;
+    }
+    // This is a flag without value, like -Zi, or -Od
+    else if (arg->getOption().getKind() == llvm::opt::Option::FlagClass) {
+      assert(arg->getNumValues() == 0);
```
Comment (llvm-beanz) [inline lib/DxcSupport/HLSLOptions.cpp:89]: nit: since the body of this condition is just an assert, we could hoist the assert out before the first `if` and have the assert contain the condition and the assert check, then this `else if` block could disappear.

**Comment by damyanp** [comment]
This PR was closed as it has not been updated in the last two years. Please feel free to reopen if this PR should be merged and is in a reviewable state.

**Comment by damyanp** [comment]
This PR was closed as it has not been updated in the last two years. Please feel free to reopen if this PR should be merged and is in a reviewable state.
